File: flang/test/Integration/array-info-delinearization-e2e.f90
! RUN: %flang_fc1 -emit-llvm %s -o - | opt -passes='print<delinearization>' -disable-output 2>&1 | FileCheck %s

! Check that delinearization successfully recognizes the 2d and 3d array structures.

! CHECK: Printing analysis 'Delinearization' for function '{{.*}}test_2d_array_access{{.*}}':
! CHECK: ArrayDecl[20][10] with elements of 4 bytes
! CHECK: ArrayRef[{{.*}}][{{.*}}]

! CHECK: Printing analysis 'Delinearization' for function '{{.*}}test_3d_array_access{{.*}}':
! CHECK: ArrayDecl[15][10][5] with elements of 4 bytes
! CHECK: ArrayRef[{{.*}}][{{.*}}][{{.*}}]

module array_test
  implicit none

  interface
    function get_index() result(idx)
      integer :: idx
    end function get_index
  end interface

contains

subroutine test_2d_array_access()
    implicit none
    integer :: arr(10, 20)
    integer :: i, j, idx, sum

    idx = get_index()
    sum = 0

    do i = 1, 5
        do j = 1, 5
            arr(i, j) = idx + i * j
            sum = sum + arr(i, j)
        end do
    end do

    do i = 1, 3
        j = mod(idx + i, 20) + 1
        arr(i, j) = sum + i
    end do

    print *, sum
    call external_use_2d(arr)
end subroutine test_2d_array_access

subroutine test_3d_array_access()
    implicit none
    integer :: arr(5, 10, 15)
    integer :: i, j, k, idx, sum

    idx = get_index()
    sum = 0

    do k = 1, 3
        do j = 1, 4
            do i = 1, 2
                arr(i, j, k) = idx + i * j * k
                sum = sum + arr(i, j, k)
            end do
        end do
    end do

    do i = 1, 2
        j = mod(idx + i, 10) + 1
        k = mod(idx + i * 2, 15) + 1
        arr(i, j, k) = sum + i * j
    end do

    print *, sum
    call external_use_3d(arr)
end subroutine test_3d_array_access

end module array_test

subroutine external_use_2d(arr)
  integer, intent(in) :: arr(10, 20)
  ! Do nothing, just prevent optimization.
end subroutine external_use_2d

subroutine external_use_3d(arr)
  integer, intent(in) :: arr(5, 10, 15)
  ! Do nothing, just prevent optimization.
end subroutine external_use_3d

File: clang/test/CodeGen/array-info-delinearization-e2e.c
// Check that LLVM IR contains the array_info assumes.
// RUN: %clang_cc1 -emit-llvm -O2 %s -o - | FileCheck %s --check-prefix=LLVM-IR
// LLVM-IR: call void @llvm.assume(i1 true) [ "array_info"(ptr %arr, i64 2, i64 10, i64 20, i64 4) ]
// LLVM-IR: call void @llvm.assume(i1 true) [ "array_info"(ptr %arr, i64 3, i64 5, i64 10, i64 15, i64 4) ]

// Test array delinearization with 2d and 3d loop nests.
// RUN: %clang_cc1 -emit-llvm -O2 %s -o - | opt -passes='print<delinearization>' -disable-output 2>&1 | FileCheck %s --check-prefix=DELINEARIZE

// DELINEARIZE: Printing analysis 'Delinearization' for function 'test_2d_loop_access':
// DELINEARIZE: Base offset: %arr
// DELINEARIZE: ArrayDecl[10][20] with elements of 4 bytes.
// DELINEARIZE: ArrayRef[{{.*}}][{{.*}}]

// DELINEARIZE: Printing analysis 'Delinearization' for function 'test_3d_loop_access':
// DELINEARIZE: Base offset: %arr
// DELINEARIZE: ArrayDecl[5][10][15] with elements of 4 bytes.
// DELINEARIZE: ArrayRef[{{.*}}][{{.*}}][{{.*}}]

extern volatile int sink;
extern void external_use(void* ptr);

void test_2d_loop_access(int n, int m) {
    int arr[10][20];

    for (int i = 0; i < n && i < 10; i++) {
        for (int j = 0; j < m && j < 20; j++) {
            arr[i][j] = i * 20 + j;
            sink = arr[i][j];
        }
    }
    external_use(arr);
}

void test_3d_loop_access(int n, int m, int p) {
    int arr[5][10][15];

    for (int i = 0; i < n && i < 5; i++) {
        for (int j = 0; j < m && j < 10; j++) {
            for (int k = 0; k < p && k < 15; k++) {
                arr[i][j][k] = i * 150 + j * 15 + k;
                sink = arr[i][j][k];
            }
        }
    }
    external_use(arr);
}
